src/main/java/org/openrewrite/java/spring/batch/MigrateStepBuilderFactory.java

```diff
@@ -0,0 +1,174 @@
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.spring.batch;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Preconditions;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaParser;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.J.ClassDeclaration;
+import org.openrewrite.java.tree.J.MethodDeclaration;
+import org.openrewrite.java.tree.Statement;
+import org.openrewrite.java.tree.TypeUtils;
+
+public class MigrateStepBuilderFactory extends Recipe {
+
+    private static final MethodMatcher STEP_BUILDER_FACTORY = new MethodMatcher(
+            "org.springframework.batch.core.configuration.annotation.StepBuilderFactory get(java.lang.String)");
+
+    @Override
+    public String getDisplayName() {
+        return "Migrate `StepBuilderFactory` to `StepBuilder`";
+    }
+
+    @Override
+    public String getDescription() {
+        return "`StepBuilderFactory` was deprecated in spring-batch 5.x. It is replaced by `StepBuilder`.";
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(5);
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        JavaVisitor<ExecutionContext> javaVisitor = new JavaVisitor<ExecutionContext>() {
+            @Override
+            public J visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                if (STEP_BUILDER_FACTORY.matches(method)) {
+
+                    ClassDeclaration clazz = getCursor().firstEnclosingOrThrow(ClassDeclaration.class);
+                    MethodDeclaration enclosingMethod = getCursor().firstEnclosingOrThrow(MethodDeclaration.class);
+
+                    maybeAddImport("org.springframework.batch.core.step.builder.StepBuilder", false);
+                    maybeRemoveImport("org.springframework.batch.core.configuration.annotation.StepBuilderFactory");
+                    maybeAddImport("org.springframework.batch.core.repository.JobRepository");
+
+                    doAfterVisit(new RemoveStepBuilderFactoryVisitor(clazz, enclosingMethod));
+
+                    JavaTemplate template = JavaTemplate
+                            .builder("new StepBuilder(#{any(java.lang.String)}, jobRepository)")
```

Comment:
From looking at the problem briefly it appears this `jobRepository` variable does not yet exist in the context when the template is applied, because that's only added as a method argument in `doAfterVisit(new RemoveStepBuilderFactoryVisitor(clazz, enclosingMethod))`. That's then likely what triggers the template error, even if the produced output would otherwise textually be ok. It might help to introduce the method argument first, before using it here.